Error categories guide (#2907)
raft-tech/TANF-app: docs/Technical-Documentation/error-categories-guide.md

> # Error Categories Guide

Reviewer: ```suggestion
# Error Categories Guide

TDP checks submitted data files (and data therein) for errors by _first_ checking if the data file being uploaded is the correct [file type](https://github.com/raft-tech/TANF-app/pull/2725), _then_ proceeding with a sequence of checks designed to determine if the data in the file can be parsed, and if so, find any issues with the data values reported. 

This guide provides a high-level overview of these types of errors. For more specific details, see [TDP's parsing directory](https://github.com/raft-tech/TANF-app/tree/develop/tdrs-backend/tdpservice/parsers) for current pre- and post-parsing validation logic.
```

the file type link i added is the PR, but could be a link to KC if that merges first.

## Category Definitions

Parity with the legacy system includes categories 1-4

> 1. **Pre-parsing errors** — These types of errors are considered to be violations of the expected record layout, so the records are not "eligible" to be parsed or validated. See [TDP's parsing directory](https://github.com/raft-tech/TANF-app/tree/develop/tdrs-backend/tdpservice/parsers) for current and complete pre-parsing logic.

Reviewer: ```suggestion
1. **Pre-parsing errors** — These types of errors are considered to be violations of the expected record layout, so the records are not "eligible" to be parsed or validated. 

- [Header and Trailer record layout](https://www.acf.hhs.gov/sites/default/files/documents/ofa/transmission_file_header_trailer_record.pdf)
- [TANF and SSP record layouts](https://www.acf.hhs.gov/sites/default/files/documents/ofa/ACF-199-%26amp%3B-209-TANF-SSP-data-report-layouts-thru-2026-10.xlsx)
- [Tribal TANF Section 1 record layout](https://www.acf.hhs.gov/sites/default/files/documents/ofa/tribal_layout_for_section1.pdf)
  - Note: Tribal TANF Sections 2 through 4 are the same as TANF Sections 2 - 4 (referenced above). 

```

> # Error Categories Guide

Reviewer: @reitermb if this is for future reference for the development team, it might be better to refer them to [this directory](https://github.com/raft-tech/TANF-app/tree/develop/tdrs-backend/tdpservice/parsers), for the most current logic around pre-parsing.  Therein, searching for `preparsing_validators` within the directory will point tech team to logic by schema and record type. 

Also, Is there a specific UX mural that tracks pre-parsing details?

Reviewer: i meant to make this comment under cat 1 😄

Author: I don't think we have a mural—or at least one that's up to date when it comes to definition of parity & error language—but I'm updating this to point to the directory!

## Category Definitions

Parity with the legacy system includes categories 1-4

1. **Pre-parsing errors** — These types of errors are considered to be violations of the expected record layout, so the records are not "eligible" to be parsed or validated. See existing reference material (in UX murals and parsing notebooks) about what these errors are and how to detect them.

   Ex:

   ```Your file does not begin with a header record```,

   ```Header length is 2 characters but must be 23```

> 2. **Out-of-range value errors –** These are based on the abovementioned instructions.

Reviewer: if we are referencing the instructions, we should link to them. they are referenced in the frontend resource cards.

   Ex:

   ```Item 76 (Citizenship/Alienage) must be in set of values [0, 1, 2, 9]```,

   ```Item 2 (County FIPS Code) must be 3 digits```

> 3. **Errors re: inconsistent values across data elements within a record –** These are also based on the abovementioned instructions (e.g. If SSI recipient = yes, then SSI amount received > $0).

Reviewer: same as above

## Plain language category names

| Category # | Internal name                                                | Plain language name      |
| ---------- | ------------------------------------------------------------ | ------------------------ |
| 1          | Pre-parsing errors                                           | File pre-check           |
| 2          | Out-of-range value errors                                    | Record value invalid     |
| 3          | Inconsistent values across data elements within a record     | Record value consistency |
| 4          | Inconsistent values across related records within a section file | Case consistency         |
| 5          | Inconsistent values across related sections of data          | Section consistency      |
| 6          | Inconsistent values across related records and/or sections over time | Historical consistency   |

## Other categorizations of TANF data

### Program types

1. TANF
2. SSP-MOE
3. Tribal TANF

### Sections

1. Active Case Data
2. Closed Case Data
3. Aggregate Data
4. Stratum (or Sample) Data

> ### Record Types

Reviewer: need to add SSP and tribal TANF record types. or at least reference the nuance.

| Record Type Indicator | Record Type Name            |
| --------------------- | --------------------------- |

> | T1                    | Family Case Characteristics |

Reviewer: ```suggestion
| T1 / M1                   | Family Case Characteristics |
```

something like this?
